addons/portal/wizard/portal_invite_share.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from odoo import api, fields, models, _
+
+
+class PortalInviteShare(models.TransientModel):
+    _name = 'portal.invite.share'
+
+    @api.model
+    def default_get(self, fields):
+        result = super(PortalInviteShare, self).default_get(fields)
+        result['res_model'] = self._context.get('active_model')
+        result['res_id'] = self._context.get('active_id')
+        result['share_link'] = self.env[result['res_model']].browse(result['res_id'])._get_share_url()
+        return result
+
+    res_model = fields.Char('Related Document Model', required=True)
+    res_id = fields.Integer('Related Document ID', required=True)
+    partner_ids = fields.Many2many('res.partner', string="Recipients", required=True)
+    note = fields.Text(help="Add extra content to display in the email")
+    share_link = fields.Char(string="Link", compute='_compute_share_link')
+    access_warning = fields.Text("Access warning", compute="_compute_share_link")
+
+    @api.depends('res_model', 'res_id')
+    def _compute_share_link(self):
+        base_url = self.env['ir.config_parameter'].sudo().get_param('web.base.url')
+        for rec in self:
+            res_model = self.env[rec.res_model]
+            if isinstance(res_model, self.pool['portal.mixin']):
+                record = res_model.browse(rec.res_id)
+                rec.access_warning = record.access_warning
+                rec.share_link = base_url + record._get_share_url()
+
+    @api.multi
+    def send_mail_action(self):
+        active_record = self.env[self.res_model].browse(self.res_id)
+        template = self.env.ref('portal.portal_share_template', False)
+        note = self.env.ref('mail.mt_note')
+        signup_enabled = self.env['ir.config_parameter'].sudo().get_param('auth_signup.invitation_scope') == 'b2c'
+
+        if hasattr(active_record, 'access_token') and active_record.access_token or not signup_enabled:
+            partner_ids = self.partner_ids
+        else:
+            partner_ids = self.partner_ids.filtered(lambda x: x.user_ids)
+        # if partner already user or record has access token send common link in batch to all user
+        for partner in self.partner_ids:
+            share_link = active_record._get_share_url(pid=partner.id)
+            active_record.with_context(mail_post_autofollow=True).message_post_with_view(template,
+                values={'name': partner.name, 'note': self.note, 'record': active_record,
```

Comment:
SMall note: I would give partner, not partner.name, to the renderer. That way if we want to customize a bit the template we have more freedom having a record than just its name.